### PR TITLE
[Experimental] Prefix Caching Support

### PR DESCRIPTION
**Pull request by caoshiyi**
add prefix caching support

Section 1 (Basic Functionality):
- [x] Test on single request for llama model (no cpu swap)
- [x] Test on batches where there are requests with prefix and without prefix (no cpu swap)
- [x] Benchmark performance for batched analytics tasks (no cpu swap)
- [x]  Alibi (Thanks @DouHappy )
- [x] Clean code

Todo:
Automatic Prefix Caching Support -- [SGLang RadixAttention](https://github.com/sgl-project/sglang)

### CODE REVIEW

**Comment by DouHappy** [comment]
It seems that the prefix has not updated its physical block?

I tested on the meta-llama/Llama-2-70b-chat-hf and baichuan2-13b-chat, but it seems to have no acceleration effect.
Then I added a logger at the 358th row of worker.py to check if prefix updated its physics block.
image

logger's output is as fellow:
------start generating------
prefix length:  592
block size:  16
Processed prompts:   0%|          | 0/500 [00:00<?, ?it/s][36m(RayWorker pid=1138291)[0m INFO 12-05 10:46:31 worker.py:358] prefix_block_tables: [[], [], [], [], []]
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:46:36 worker.py:358] prefix_block_tables: [[], [], [], [], []][32m [repeated 44x across cluster][0m
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:46:42 worker.py:358] prefix_block_tables: [[], [], [], [], []][32m [repeated 44x across cluster][0m
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:46:47 worker.py:358] prefix_block_tables: [[], [], [], [], []][32m [repeated 44x across cluster][0m
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:46:52 worker.py:358] prefix_block_tables: [[], [], [], []]
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:46:52 worker.py:358] prefix_block_tables: []
[36m(RayWorker pid=1138290)[0m INFO 12-05 10:46:51 worker.py:358] prefix_block_tables: [[], [], [], [], []][32m [repeated 39x across cluster][0m
Processed prompts:   0%|          | 1/500 [00:23<3:18:23, 23.86s/it][36m(RayWorker pid=1138290)[0m INFO 12-05 10:46:52 worker.py:358] prefix_block_tables: [[], [], [], []][32m [repeated 3x across cluster][0m
[36m(RayWorker pid=1138290)[0m INFO 12-05 10:46:55 worker.py:358] prefix_block_tables: [][32m [repeated 59x across cluster][0m
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:46:57 worker.py:358] prefix_block_tables: [[], [], [], [], []][32m [repeated 21x across cluster][0m
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:47:02 worker.py:358] prefix_block_tables: [[], [], [], [], []][32m [repeated 44x across cluster][0m
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:47:08 worker.py:358] prefix_block_tables: [[], [], [], [], []][32m [repeated 44x across cluster][0m
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:47:13 worker.py:358] prefix_block_tables: [[], [], [], [], []][32m [repeated 44x across cluster][0m
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:47:15 worker.py:358] prefix_block_tables: [[], [], []]
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:47:16 worker.py:358] prefix_block_tables: []
Processed prompts:  43%|████▎     | 217/500 [00:45<00:50,  5.65it/s][36m(RayWorker pid=1138291)[0m INFO 12-05 10:47:16 worker.py:358] prefix_block_tables: [[], []]
Processed prompts:  44%|████▍     | 219/500 [00:45<00:49,  5.65it/s][36m(RayWorker pid=1138291)[0m INFO 12-05 10:47:17 worker.py:358] prefix_block_tables: [[]]
[36m(RayWorker pid=1138290)[0m INFO 12-05 10:47:15 worker.py:358] prefix_block_tables: [[], [], [], [], []][32m [repeated 19x across cluster][0m
Processed prompts:  44%|████▍     | 220/500 [00:47<00:53,  5.24it/s][36m(RayWorker pid=1138290)[0m INFO 12-05 10:47:15 worker.py:358] prefix_block_tables: [[], [], []][32m [repeated 3x across cluster][0m
[36m(RayWorker pid=1138290)[0m INFO 12-05 10:47:18 worker.py:358] prefix_block_tables: [][32m [repeated 59x across cluster][0m
[36m(RayWorker pid=1138290)[0m INFO 12-05 10:47:16 worker.py:358] prefix_block_tables: [[], []][32m [repeated 3x across cluster][0m
[36m(RayWorker pid=1138290)[0m INFO 12-05 10:47:17 worker.py:358] prefix_block_tables: [[]][32m [repeated 3x across cluster][0m
[36m(RayWorker pid=1138291)[0m INFO 12-05 10:47:23 worker.py:358] prefix_block_tables: [[], [], [], [], []][32m [repeated 41x across cluster][0m
Processed prompts: 100%|██████████| 500/500 [00:54<00:00,  9.13it/s]cost time 56.1341028213501
saving output

[36m(RayWorker pid=1138291)[0m INFO 12-05 10:47:26 worker.py:358] prefix_block_tables: []


image

By the way, My prompts are generated by this script:
```
# generate test prompt
test_table = "|Opening|Opening|Sl. No.|Film|Cast|Director|Music Director|Notes|\n|----|----|----|----|----|----|----|----|\n|J A N|9|1|Agni Pushpam|Jayabharathi, Kamalahasan|Jeassy|M. K. Arjunan||\n|J A N|16|2|Priyamvada|Mohan Sharma, Lakshmi, KPAC Lalitha|K. S. Sethumadhavan|V. Dakshinamoorthy||\n|J A N|23|3|Yakshagaanam|Madhu, Sheela|Sheela|M. S. Viswanathan||\n|J A N|30|4|Paalkkadal|Sheela, Sharada|T. K. Prasad|A. T. Ummer||\n|F E B|5|5|Amma|Madhu, Srividya|M. Krishnan Nair|M. K. Arjunan||\n|F E B|13|6|Appooppan|Thikkurissi Sukumaran Nair, Kamal Haasan|P. Bhaskaran|M. S. Baburaj||\n|F E B|20|7|Srishti|Chowalloor Krishnankutty, Ravi Alummoodu|K. T. Muhammad|M. S. Baburaj||\n|F E B|20|8|Vanadevatha|Prem Nazir, Madhubala|Yusufali Kechery|G. Devarajan||\n|F E B|27|9|Samasya|Madhu, Kamalahaasan|K. Thankappan|Shyam||\n|F E B|27|10|Yudhabhoomi|K. P. Ummer, Vidhubala|Crossbelt Mani|R. K. Shekhar||\n|M A R|5|11|Seemantha Puthran|Prem Nazir, Jayabharathi|A. B. Raj|M. K. Arjunan||\n|M A R|12|12|Swapnadanam|Rani Chandra, Dr. Mohandas|K. G. George|Bhaskar Chandavarkar||\n|M A R|19|13|Thulavarsham|Prem Nazir, sreedevi, Sudheer|N. Sankaran Nair|V. Dakshinamoorthy||\n|M A R|20|14|Aruthu|Kaviyoor Ponnamma, Kamalahasan|Ravi|G. Devarajan||\n|M A R|26|15|Swimming Pool|Kamal Haasan, M. G. Soman|J. Sasikumar|M. K. Arjunan||"

prompt_template = '''\
You are a helpful assistant in recongnizes the content of tables in markdown format. Here is a table as fellows. You need to answer my question about the table.
# Table
{}

# Question
What' s the content in the ({},{}) cells
'''
with open('prompt.txt', 'w') as outer:
    for row in range(50):
        for column in range(10):
            tmp_str = prompt_template.format(test_table, row + 1, column + 1)
            tmp_str
            # outer.write(f"{tmp_str}")
            print(tmp_str.replace("\n", "\\n"), file=outer)
```

Maybe there are some bugs in my test?

**Comment by caoshiyi** [comment]
@DouHappy Can you try with calling `llm.generate()` with one prompt first to warmup? `prefix_block_tables=[[],[],[]]` indicates that the kv cache for the prefix part hasn't been computed yet.

**Comment by DouHappy** [comment]
> @DouHappy Can you try with calling `llm.generate()` with one prompt first to warmup? `prefix_block_tables=[[],[],[]]` indicates that the kv cache for the prefix part hasn't been computed yet. Also, can you share your testing script?

I find a bug that when using two GPUs is slower than single GPU. Prefix‘s state 'on_gpu' is always False before prepare_inputmatedata() When using two GPUs. and it works nice on single gpu. It mean multi_query_cached_kv_attention never be used when running on multi-gpus. My last test is also pass on single gup.

File: vllm/worker/worker.py [hunk 1]
```diff
@@ -208,24 +225,29 @@ def _prepare_inputs(
...
+                prompt_len = subquery_lens[i]
                 if sampling_params.prompt_logprobs is not None:
                     selected_token_indices.extend(
                         range(selected_token_start_idx,
                               selected_token_start_idx + prompt_len - 1))
                 selected_token_indices.append(selected_token_start_idx +
                                               prompt_len - 1)
                 selected_token_start_idx += max_seq_len
+
+                # set the prefix state
```
Comment (yudian0504) [inline vllm/worker/worker.py:247]: when tp>1, seq_group_metadata.prefix here is copied  by ray workers, so  `on_gpu=true` won't work on multi gpus.

Comment (yudian0504) [reply vllm/worker/worker.py:247]: cc. @DouHappy

Comment (DouHappy) [reply vllm/worker/worker.py:247]: Thank you for your reply. This help me a lot.

File: vllm/prefix.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        token_ids: List[int],
+        block_size: int,
+    ) -> None:
+        self.prefix_id = prefix_id
+        self.token_ids = token_ids
+        self.length = len(token_ids)
+        assert self.length % block_size == 0
+        self.on_gpu = False
+        self.on_cpu = False
+        self.block_table = None
```
Comment (zhuohan123) [inline vllm/prefix.py:30]:
```suggestion
        self.block_table: Optional[List[int]] = None
```

Comment (zhuohan123) [reply vllm/prefix.py:30]: BTW, when will this field being set?

File: vllm/prefix.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        self.block_size = block_size
+
+    def add_prefix(self, token_ids: List[int]) -> Prefix:
+        # generate prefix_id
+        prefix_id = len(self.prefixes)
+        # create a new prefix
+        prefix = Prefix(prefix_id, token_ids, self.block_size)
+        self.prefixes.append(prefix)
+        prefix_hash = hash(tuple(prefix.token_ids))
+        self.prefixes_hash[prefix_hash] = prefix.prefix_id
```
Comment (zhuohan123) [inline vllm/prefix.py:75]: Assert that the prefix has not been added yet?

File: vllm/prefix.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        self.token_ids = token_ids
+        self.length = len(token_ids)
+        assert self.length % block_size == 0
+        self.on_gpu = False
+        self.on_cpu = False
+        self.block_table = None
+        # a lock to prevent multiple sequence from calculating the same prefix
+        self.swap_to_gpu = False
+
+    def get_block_table_num(self) -> List[int]:
```
Comment (zhuohan123) [inline vllm/prefix.py:34]:
```suggestion
    def get_block_numbers(self) -> List[int]:
```

File: vllm/prefix.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        on_gpu: True if the prefix will be on GPU before the execution of the model.
+        on_cpu: True if the prefix is on CPU.
+        swap_to_gpu: True when the prefix will be computed during the execution of the model.
+    """
+
+    def __init__(
+        self,
+        prefix_id: int,
+        token_ids: List[int],
+        block_size: int,
```
Comment (zhuohan123) [inline vllm/prefix.py:22]: Seems like `block_size` is not stored in the class.

**Review by zhuohan123 — CHANGES_REQUESTED** [review]
Thanks for the great work! Can you also merge with the latest main branch as well? I will test the PR after the merge.

File: vllm/model_executor/layers/triton_kernel/prefix_prefill.py [hunk 6]
```diff
@@ -0,0 +1,910 @@
...
+                (alibi <= 0) & (alibi_start_q[:, None] < cur_batch_seq_len),
+                alibi, float("-inf"))
+            qk += alibi
+            alibi_start_k += BLOCK_N
+
+            # # debuger alibi
+            # offset_db = stride_db_head * cur_head + offset_db_q[:, None] * stride_db_q + offset_db_k[None, :] * stride_db_k
+            # mask_db = (offset_db_q < cur_batch_seq_len - cur_batch_ctx_len)[:, None] & (offset_db_k < cur_batch_seq_len)[None, :]
+            # tl.store(debuger + offset_db, alibi, mask=mask_db)
+            # offset_db_k += BLOCK_N
```
[the remark this answers is not in the document]
Comment (DouHappy) [reply vllm/model_executor/layers/triton_kernel/prefix_prefill.py:587]: Sorry, These are commented-out code. Will be removed.

File: vllm/engine/async_llm_engine.py [hunk 7]
```diff
@@ -400,13 +402,15 @@ async def add_request(
...
-            arrival_time=arrival_time)
+            arrival_time=arrival_time,
+            prefix_pos=prefix_pos)
 
         return stream
 
     async def generate(
         self,
         prompt: Optional[str],
+        prefix_pos: Optional[int],
```
Comment (chenxu2048) [inline vllm/engine/async_llm_engine.py:413]: Hi, @caoshiyi.

This function might be used by others who use the engine in their project directly.
To ensure backward compatibility, I thought it is better to add new parameters at the end, just like `AsyncLLMEngine.add_request` and `LLMEngine.add_request`.

File: vllm/model_executor/input_metadata.py [hunk 8]
```diff
@@ -18,14 +18,21 @@ def __init__(
         self,
         prompt_lens: List[int],
         slot_mapping: torch.Tensor,
+        max_seq_len: Optional[int],
```
[the remark this answers is not in the document]
Comment (caoshiyi) [reply vllm/model_executor/input_metadata.py:22]: no, they are max of subquery_lens.

File: vllm/prefix.py (hunk 2 above)
Comment (caoshiyi) [reply vllm/prefix.py:30]: It is set the block_manager.py: allocate().

**Comment by jadielam** [comment]
I propose the changes above to the `PrefixPool` class.
1. There is no need to maintain two data structures: one being a dict from `prefix_hash` to `prefix_id` and the other one being the list of prefixes.  We can simply maintain just the dictionary, directly from `prefix_hash` to `Prefix`.

2. I think that the prefix pool class should have a max_capacity, otherwise it can have the potential of growing without bounds. Maybe in your mind, the prefix use case does not need it. But we have some use cases where we are planning to use prefixes differently and there will be an unlimited number of them. 
2.1 Because of that I propose to implement it using an OrderedDict so that we can easily implement a FIFO cache policy.

```
class PrefixPool:
    """Manages all the prompt prefixes.

    Args:
        block_size: The block size of the executed model.
    
    Attributes:
        prefixes: A list of all the prefixes.
        prefixes_hash: Mapping from the hash of the prefix to the prefix id.
        block_size: The block size of the executed model.
        max_capacity: Max number of prefixes to keep in the pool at any given time.
    """
    id_iter = itertools.count()

    def __init__(
        self,
        block_size: int,
        max_capacity: int = 32
    ) -> None:
        # dict from hash to prefix. It is an OrderedDict so that
        # we can easily implement a FIFO cache policy.
        self.prefixes: dict[int, Prefix] = OrderedDict() 
        
        self.block_size = block_size    # Number of tokens per memory block
        self.max_capacity=max_capacity

    def add_prefix(self, token_ids: list[int]) -> Prefix:
        if len(self.prefixes) >= self.max_capacity:
            self.prefixes.popitem(last=False)

        prefix_hash = hash(tuple(token_ids))
        assert prefix_hash not in self.prefixes
        
        # generate prefix_id and create new prefix
        prefix_id = next(self.id_iter)
        prefix = Prefix(prefix_id, token_ids, self.block_size)
        
        self.prefixes[prefix_hash] = prefix
        return prefix

    # use this first, if we already know from the application which part of the tokens are prefix.
    def fixed_search(self, prefix_hash: int) -> Optional[Prefix]:
        return self.prefixes.get(prefix_hash, None)
```

File: vllm/prefix.py [hunk 9]
```diff
@@ -1,88 +1,79 @@
...
+    def add_or_get_prefix(self, token_ids: Sequence[int]) -> Optional[Prefix]:
+        token_ids = self._truncate_token_ids(token_ids)
+        if len(token_ids) == 0:
+            # Prefix is empty.
             return None
-        prefix_id = self.prefixes_hash[prefix_hash]
-        return self.prefixes[prefix_id]
+        prefix = Prefix(token_ids, self.block_size)
+        prefix_hash = prefix.get_hash()
+        if prefix_hash not in self.prefixes_hash:
```
Comment (jadielam) [inline vllm/prefix.py:77]: `self.prefixes_hash` no longer exists.

File: vllm/prefix.py [hunk 10]
```diff
@@ -1,88 +1,79 @@
...
+
+    def add_or_get_prefix(self, token_ids: Sequence[int]) -> Optional[Prefix]:
+        token_ids = self._truncate_token_ids(token_ids)
+        if len(token_ids) == 0:
+            # Prefix is empty.
             return None
-        prefix_id = self.prefixes_hash[prefix_hash]
-        return self.prefixes[prefix_id]
+        prefix = Prefix(token_ids, self.block_size)
+        prefix_hash = prefix.get_hash()
```
Comment (jadielam) [inline vllm/prefix.py:76]: You could implement the `__hash__` function in the `Prefix` class and by that eliminate some steps here.

**Comment by jadielam** [comment]
- I am interested in contributing to this branch. I have opened a PR with some changes to it. Feel free to review it. https://github.com/caoshiyi/vllm/pull/2

- The changes introduced in that PR do not really effectively add new functionality to the vllm engine right now, but they are stepping stones to make it happen in a later PR. 

- The idea that I have in mind is the following: Whenever the processing of a `SequenceGroup` is either finished, or aborted, check the `prefix_pool: PrefixPool` object and see if the `prefix` (if any) of that `SequenceGroup` still exists in the `PrefixPool`. If it does not exists, then free the GPU memory from that prefix.  

I have read the code base extensively and I don't think that adding this new functionality affects the behavior of the system as intended originally, given that by default `PrefixPool` will behave as if no `max_capacity` has been set.

**Comment by zhuohan123** [comment]
@jadielam Thank you for your contribution! How about the following plan: I will fix the bugs of the PR and first merge a version without prefix pool capacity, and then you create a new PR in the vLLM repo to add `max_capacity` to prefix pool?

**Comment by jadielam** [comment]
> he vLLM repo to add `max_ca`

Sounds good.

**Comment by jadielam** [comment]
> @jadielam Thank you for your contribution! How about the following plan: I will fix the bugs of the PR and first merge a version without prefix pool capacity, and then you create a new PR in the vLLM repo to add `max_capacity` to prefix pool?

sounds good.

**Comment by franklyd** [comment]
Thanks a lot for this great feature! 
I tried it with the latest caoshiyi:prefix, but I found that there's no speed improvement. (one V100 GPU, tested with Baichuan2-13B-chat model)

Hi @DouHappy , did you observe any speed improvement afterwards?

**Comment by jadielam** [comment]
> Thanks a lot for this great feature! I tried it with the latest caoshiyi:prefix, but I found that there's no speed improvement. (one V100 GPU, tested with Baichuan2-13B-chat model)
> 
> Hi @DouHappy , did you observe any speed improvement afterwards?

Hi @franklyd I also haven't tested. Have you tested regarding throughput improvements? This is another variable that should see improvements (under certain circumstances) from this PR.

**Comment by zhuohan123** [comment]
@franklyd @DouHappy There was a bug in my refactor. If you try now, you should be able to see speedups.

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thanks for the great work! Pushed some style refactors by myself.

**Review by HaiShaw — COMMENTED** [review]
@zhuohan123 

I think this merge now invalidates FP8 KV cache (#2279).
Look at `kernels` in `prefix_prefill.py`, when FP8 KV cache is `ON`, `K/V` and `K_cache/V_cache` are different types now.

Please let know what is the best way to move forward, thanks!

**Comment by gangooteli** [comment]
> > Thanks a lot for this great feature! I tried it with the latest caoshiyi:prefix, but I found that there's no speed improvement. (one V100 GPU, tested with Baichuan2-13B-chat model)
> > Hi @DouHappy , did you observe any speed improvement afterwards?
> 
> Yes，i got observe speed up. Could you should me your test script? Maybe you forgot warmup? BTW, I am trying to introduce prefix but only chinese version now. See this [vLLM-prefix浅析（System Prompt，大模型推理加速）](https://zhuanlan.zhihu.com/p/678256296) @franklyd

Could you provide a test script for the speedup?

**Comment by ksjadeja** [comment]
> Could you provide a test script for the speedup?

+1

**Comment by AlpinDale** [comment]
Hi @HaiShaw 

Triton doesn't seem to support mixed precision dot product, so this kernel here fails if the `k` is uint8 and `q` is another precision. I've been trying to find a solution to this problem, but coming up with blanks. Do you have any ideas on how to approach this?

**Comment by chenxu2048** [comment]
> Hi @HaiShaw
> 
> Triton doesn't seem to support mixed precision dot product, so this kernel here fails if the `k` is uint8 and `q` is another precision. I've been trying to find a solution to this problem, but coming up with blanks. Do you have any ideas on how to approach this?

Hi, @AlpinDale. Are you using prefix caching with FP8 KVCache? PyTorch and Triton used by vLLM could not support FP8 KVCache. Here are more information about prefix caching and FP8 KVCache in #3234.

**Comment by Daya-Jin** [comment]
brilliant feature, thx!

**Pull request merged**